[BE] Public/private job

MihaiDruta opened the pull request:
[BE] Public/private job
https://trello.com/c/0yC8EhYU/42-be-public-private-job

Job could be private/public. Implemented function to invite a user to a job. Users could see only available activities in activity list.

Commit:
[BE] Public/private job
https://trello.com/c/0yC8EhYU/42-be-public-private-job

Job could be private/public. Implemented function to invite a user to a job. Users could see only available activities in activity list.

SpartakusMd commented on src/Entity/Activity.php:
```diff
@@ -300,4 +315,14 @@ public function updatedTimestamps(): void
+        return $this->public;
+    }
+
+    public function setPublic($public): void
```
`setPublic(bool $public)`

SpartakusMd commented on src/Migrations/Version20190619140837.php:
```diff
@@ -0,0 +1,37 @@
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql',
+            'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE activity ADD privacy TINYINT(1) NOT NULL');
```
`public` not `privacy`.

SpartakusMd replied:
Actualy I saw next migration in which you do the renaming. Remove both of them so you have one migration which is correct from the start.

SpartakusMd commented on src/Security/AccessRightsPolicy.php:
```diff
@@ -0,0 +1,31 @@
+     */
+    private $repository;
+
+    public function __construct(ActivityUserRepository $repository)
```
Use full variable name `$activityUserRepository` so later you know for which entity is this repository without scrolling to the constructor.

SpartakusMd commented on src/Security/AccessRightsPolicy.php:
```diff
@@ -0,0 +1,31 @@
+        }
+        return true;
+    }
+}
```
Add empty line at the end.

SpartakusMd commented on src/Security/AccessRightsPolicy.php:
```diff
@@ -0,0 +1,31 @@
+
+    public function checkRightsToActivity(Activity $activity, User $user): bool
+    {
+        $userToActivity = $this->repository->findBy(array('activity' => $activity, 'user' => $user));
```
Do the check `$activity->isPublic() === false) && $user !== $activity->getOwner()` before checking if user is invited/has applied. Most of the jobs will be public and the request to the database won't be needed.

SpartakusMd commented on src/DTO/ActivityDTO.php:
```diff
@@ -88,6 +87,22 @@ class ActivityDTO
+
+    /**
+     * @var boolean
+     * @Serializer\Type("integer")
```
Type `boolean` maybe? https://jmsyst.com/libs/serializer/master/reference/annotations#type

Commit:
[BE] Public/private job
https://trello.com/c/0yC8EhYU/42-be-public-private-job

Job could be private/public. Implemented function to invite a user to a job. Users could see only available activities in activity list.

SpartakusMd commented on src/Security/AccessRightsPolicy.php:
```diff
@@ -14,18 +14,20 @@ class AccessRightsPolicy
 
-        if (empty($userToActivity) && ($activity->isPublic() === false) && $user !== $activity->getOwner()) {
-            return false;
+        if (($activity->isPublic() === false) && $user !== $activity->getOwner()) {
```
Use early return. You should always avoid nesting if statements.

Commit:
[BE] Public/private job
https://trello.com/c/0yC8EhYU/42-be-public-private-job

Job could be private/public. Implemented function to invite a user to a job. Users could see only available activities in activity list.

SpartakusMd commented on src/DTO/ActivityDTO.php:
```diff
@@ -88,6 +87,22 @@ class ActivityDTO
+    /**
+     * @var boolean
+     * @Serializer\Type("boolean")
+     * @Assert\Range(
```
This should not be required and should not work.

SpartakusMd commented on src/Security/AccessRightsPolicy.php:
```diff
@@ -0,0 +1,31 @@
+        $this->repository = $activityUserRepository;
+    }
+
+    public function checkRightsToActivity(Activity $activity, User $user): bool
```
Rename to `canAccessActivity`

Commit:
[BE] Public/private job
https://trello.com/c/0yC8EhYU/42-be-public-private-job

Job could be private/public. Implemented function to invite a user to a job. Users could see only available activities in activity list.
